api: pass user object to auth hooks

joselcvarela opened the pull request:
The motivation for this is the ability to migrate from
bcrypt passwords to argon2 seamlessly.

Fixes #4718
Fixes #4335

joselcvarela committed:
api: pass user object to auth hooks
The motivation for this is the ability to migrate from
bcrypt passwords to argon2 seamlessly.

Fixes #4718
Fixes #4335

joselcvarela commented:
This made possible to migrate from any system to directus.
In my use case, we could not send an reset link email to users.

In #4335, @rijkvanzanten mentioned this could be a vulnerability. But we can also break all the system with this hook https://github.com/directus/directus/blob/7262fafffaef55f20891fe67db9c3159e75996c6/api/src/app.ts#L182 since we have access to app.

Also everyone should be responsible for what they are putting on their system and I don't think `directus` should be so restrictive, otherwise developers can't do much.

Let me hear your thoughts.

rijkvanzanten commented on api/src/services/authentication.ts:
```diff
@@ -74,6 +74,7 @@ export class AuthenticationService {
 			accountability: this.accountability,
 			status: 'pending',
 			user: user?.id,
+			obj: user,
```
Instead of a new `obj` property, lets just not omit stuff from the `hookPayload` 👍🏻

joselcvarela replied:
The intent of use `obj` and pass `user` is to modify directly the `user.password` property so it will not fail here:
https://github.com/directus/directus/blob/13db4c77193970614c20be09eb6f36726296f8f1/api/src/services/authentication.ts#L128

rijkvanzanten replied:
Oh we shouldn't do that mutable! Instead, lets catch the altered user payload coming from the emitter and use that to login with

joselcvarela committed:
receive user from `auth.login.before` hook

joselcvarela commented:
@rijkvanzanten something like this 93b2235?

rijkvanzanten commented:
Exactly like that 👍🏻 Do we still want to omit the otp value?

joselcvarela committed:
pass `otp` as well on `auth.login` hooks

joselcvarela commented:
Since we are allowing password, it makes sense to pass OTP too. Someone will have an use case for this 😄

joselcvarela commented:
Can I do something in order to merge this?

rijkvanzanten commented on api/src/services/authentication.ts:
```diff
@@ -58,31 +57,33 @@ export class AuthenticationService {
 			.first();
 
-		await emitter.emitAsync('auth.login.before', hookPayload, {
+		const updatedUser = await emitter.emitAsync('auth.login.before', options, {
```
Last thing: We'll have to make sure this returns the correct thing. IIRC the emitter returns an array of payloads, as multiple hooks can modify the same payload in order of registration. We can use the following to merge those back into the final augmented object:

```js
const newPayload = hooksResult.reverse().reduce((val, acc) => merge(acc, val))
```

joselcvarela replied:
I am making some tests locally and I think we should not use the `reverse` because I think the last registered hook should have priority on changes. This means on merge, last hook should have priority over others. Don't you agree?

joselcvarela replied:
I ended up using only the `reduce` since I think it makes more sense. But we can replace with `reduceRight` if you think it's better.

rijkvanzanten replied:
I'll have to double check why it is like that in the first place. Regardless, I wanna keep it the same in both places, so if it makes most sense to drop `reverse`, we should do it everywhere 👍🏻

joselcvarela replied:
So, should I put `reverse` in this PR and open another which removes `reverse` from hook handling?
I am only seeing the other cases here:
https://github.com/directus/directus/blob/31113a3087973237f52883199afeaed4540713c7/api/src/services/items.ts#L96
https://github.com/directus/directus/blob/31113a3087973237f52883199afeaed4540713c7/api/src/services/items.ts#L375
https://github.com/directus/directus/blob/31113a3087973237f52883199afeaed4540713c7/api/src/controllers/not-found.ts#L19

rijkvanzanten replied:
We can do it all in this PR, that's fine. I just want to make sure it's consistent everywhere, and the expected behavior everywhere 👍🏻

joselcvarela replied:
No problem. Here it is 5885b69ecd29ca2a4e94e00a7a4a1f912c6e76a2 ;)

joselcvarela committed:
api: last registered hook should have priority
When registering multiple hooks for a single event,
the last hook should have priority on return value.

joselcvarela commented:
Can I do something in order to merge this?

rijkvanzanten commented:
@joselcvarela Nope, just a bit of patience 😄 It's on my to-do list 👍🏻

joselcvarela commented:
Sorry @rijkvanzanten I was in a hurry to have this for our client. But I figured another way out😉
Although, I am still available to do some improvements if needed.

(I just realized I have written the same comment between 14 days without seeing it 😲)

rijkvanzanten commented:
Thanks! Little hard to prioritize specific things on user needs if there's 70/30 other issues/prs that are important to others as well 🙂

rijkvanzanten committed:
Merge branch 'main' into pass-user-auth-hooks

rijkvanzanten committed:
Fix linter warning
